beacon_node/network/src/subnet_service/data_column_subnets.rs

```diff
@@ -0,0 +1,311 @@
+//! This service keeps track of which data column subnets the beacon node should be subscribed to at any
+//! given time. It schedules subscriptions to data column subnets and requests peer discoveries.
+
+use futures::prelude::*;
+use std::task::{Context, Poll};
+use std::{
+    collections::{HashMap, VecDeque},
+    pin::Pin,
+    sync::Arc,
+    time::Duration,
+};
+
+use beacon_chain::{BeaconChain, BeaconChainTypes};
+use delay_map::HashSetDelay;
+use lighthouse_network::{discovery::peer_id_to_node_id, NetworkGlobals, Subnet, SubnetDiscovery};
+use slog::{debug, o, trace, warn, error};
+use slot_clock::SlotClock;
+use types::{ChainSpec, DataColumnSubnetId, Epoch, EthSpec};
+
+use super::SubnetServiceMessage;
+
+// The minimum number of slots ahead that we attempt to discover peers for a subscription. If the
+/// slot is less than this number, skip the peer discovery process.
+/// Subnet discovery query takes at most 30 secs, 2 slots take 24s.
+const MIN_PEER_DISCOVERY_SLOT_LOOK_AHEAD: u64 = 2;
+
+/// A particular subnet at a given slot.
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct ExactSubnet {
+    /// The `DataColumnSubnetId` associated with this subnet.
+    pub data_column_subnet_id: DataColumnSubnetId,
+    /// The epoch until which we need to stay subscribed to the subnet.
+    pub until_epoch: Epoch,
+}
+
+pub struct DataColumnService<T: BeaconChainTypes> {
+    /// Queued events to return to the driving service.
+    events: VecDeque<SubnetServiceMessage>,
+
+    /// A reference to the beacon chain to process data columns.
+    pub(crate) beacon_chain: Arc<BeaconChain<T>>,
+
+    /// A reference to the nodes network globals
+    network_globals: Arc<NetworkGlobals<T::EthSpec>>,
+
+    /// The collection of all currently subscribed data column subnets by epoch.
+    subscriptions: HashMap<DataColumnSubnetId, Epoch>,
+
+    /// A collection of timeouts for when to unsubscribe from a subnet.
+    unsubscriptions: HashSetDelay<DataColumnSubnetId>,
+
+    /// The waker for the current thread.
+    waker: Option<std::task::Waker>,
+
+    chain_spec: ChainSpec,
+
+    /// The logger for the data column service.
+    log: slog::Logger,
+}
+
+impl<T: BeaconChainTypes> DataColumnService<T> {
```

Comment:
I think we'd also need something like [`recompute_long_lived_subnets`](https://github.com/sigp/lighthouse/blob/84a902a5890e0354aa464d73ed1c3ef669d2ad10/beacon_node/network/src/subnet_service/attestation_subnets.rs#L284) here.

Reply:
Added similar logic